internal/consoleplugin/reconciler.go

```diff
@@ -0,0 +1,170 @@
+	service      *corev1.Service
+	deployment   *appv1.Deployment
+	ownedObjects []client.Object
+}
+
+// NewReconciler uses client to reconcile to the state required for config.
+func NewReconciler(c client.Client, cf *Config) *Reconciler {
+	r := &Reconciler{
+		c:             c,
+		createdBy:     cf.CreatedBy,
+		consolePlugin: cf.consolePlugin(),
+		configMap:     cf.configMap(),
+		service:       cf.service(),
+		deployment:    cf.deployment(),
+	}
+	r.ownedObjects = []client.Object{
+		r.configMap,
+		r.service,
+		r.deployment,
+	}
+	return r
+}
+
+// Reconcile create or update cluster objects to match config.
+// Returns AlreadyExists if a consoleplugin exists with the same name and different created-by label.
+func (r *Reconciler) Reconcile(ctx context.Context) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("error reconciling logging console plugin: %w", err)
+		}
+	}()
+	if err = r.reconcileConsolePlugin(ctx); err != nil {
+		return err
+	}
+	// Set consolePlugin as controller-owner for all other objects
+	for _, o := range r.ownedObjects {
+		if err = controllerutil.SetControllerReference(r.consolePlugin, o, r.c.Scheme()); err != nil {
+			return err
+		}
+	}
+	// Reconcile owned objects
+	for _, f := range []func(ctx context.Context) error{
+		r.reconcileConfigMap,
+		r.reconcileService,
+		r.reconcileDeployment,
+	} {
+		if err := f(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileConsolePlugin(ctx context.Context) error {
+	want := r.consolePlugin
+	got := &consolev1alpha1.ConsolePlugin{}
+	return r.reconcileObject(ctx, want, got, func() (bool, error) {
+		if got.Labels == nil || got.Labels[labelAppK8sCreatedBy] != r.createdBy {
+			// ConsolePlugin exists but was not created by us. Leave it alone.
+			gvk, _ := apiutil.GVKForObject(want, r.c.Scheme())
+			rm, _ := r.c.RESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version)
+			return false, errors.NewAlreadyExists(rm.Resource.GroupResource(), want.Name)
+		}
+		if reflect.DeepEqual(want.Spec, got.Spec) {
+			// No update, save the actual state. Need UID for owner reference.
+			got.DeepCopyInto(want)
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+func (r *Reconciler) reconcileConfigMap(ctx context.Context) error {
+	want := r.configMap
+	got := &corev1.ConfigMap{}
+	return r.reconcileObject(ctx, want, got, func() (bool, error) {
+		return !reflect.DeepEqual(want.Data, got.Data), nil
+	})
+}
+
+func (r *Reconciler) reconcileDeployment(ctx context.Context) error {
+	want := r.deployment
+	got := &appv1.Deployment{}
+	return r.reconcileObject(ctx, want, got, func() (bool, error) {
+		return !reflect.DeepEqual(want.Spec, got.Spec), nil
+	})
+}
+
+func (r *Reconciler) reconcileService(ctx context.Context) error {
+	want := r.service
+	got := &corev1.Service{}
+	return r.reconcileObject(ctx, want, got, func() (bool, error) {
+		if reflect.DeepEqual(want.Spec.Ports, got.Spec.Ports) &&
```

Comment:
We likely already have a service reconciliation that does this

Reply:
where would I look for that?

Reply:
here but doesnt look generic.  https://github.com/openshift/cluster-logging-operator/blob/master/internal/k8shandler/collection.go#L197  I know EO has made many of these reconcilers generic

Reply:
Can't find it, e.g. the collector service is reconciled directly:  https://github.com/alanconway/cluster-logging-operator/blob/logging-view-plugin/internal/k8shandler/collection.go#L197